src/misc.ts

```diff
@@ -92,3 +92,87 @@ export const hasProperty = (
   object: RuntimeObject,
   name: string | number | symbol,
 ): boolean => Object.hasOwnProperty.call(object, name);
+
+export type PlainObject = Record<number | string | symbol, unknown>;
+
+/**
+ * Predefined sizes (in Bytes) of specific parts of JSON structure.
+ */
+export enum JsonSize {
+  NULL = 4,
+  COMMA = 1,
+  WRAPPER = 1,
+  TRUE = 4,
+  FALSE = 5,
+  ZERO = 1,
+  QUOTE = 1,
+  COLON = 1,
+  DOT = 1,
+}
+
+/**
+ * Check if the value is plain object.
+ *
+ * @param value - Value to be checked.
+ * @returns Boolean.
+ */
+export function isPlainObject(value: unknown): value is PlainObject {
+  if (typeof value !== 'object' || value === null) {
+    return false;
+  }
+
+  try {
+    let proto = value;
+    while (Object.getPrototypeOf(proto) !== null) {
+      proto = Object.getPrototypeOf(proto);
+    }
+
+    return Object.getPrototypeOf(value) === proto;
+  } catch (_) {
+    return false;
+  }
+}
+
+/**
+ * Check if character or string is ASCII.
+ *
+ * @param value - String or character.
+ * @returns Boolean, true if value is ASCII, false if not.
+ */
+export function isASCII(value: string) {
+  // eslint-disable-next-line no-control-regex,require-unicode-regexp
+  return /^[\x00-\x7F]*$/.test(value);
+}
+
+/**
+ * Calculate string size.
+ *
+ * @param value - String value to calculate size.
+ * @returns Number of bytes used to store whole string value.
+ */
+export function calculateStringSize(value: string) {
+  let size = 0;
+  for (const character of value) {
+    if (isASCII(character)) {
+      size += 1;
+    } else {
+      size += 2;
+    }
+  }
+
+  // Detect characters that need backslash escape
+  const re = /"|\\|\n|\r|\t/gu;
+  size += (value.match(re) || []).length;
+
+  return size;
+}
+
+/**
+ * Calculate size of a number ofter JSON serialization.
+ *
+ * @param value - Number value to calculate size.
+ * @returns Number of bytes used to store whole number in JSON.
+ */
+export function calculateNumberSize(value: number): number {
+  return value.toString().length;
```

Comment:
It doesn't work for for many numbers actually in that way. I had to do serialization at the end. 
The problem with with that logarithmic approach is that it cannot handle all the possible numbers and notations.
I had really robust function for this and at the end I had to drop the solution because it cannot support exponential notation (e.g. -5e-11, etc.).

Reply:
So, to elaborate on this.
These are two functions that I previously used to calculate size of a numbers:

```
/**
 * Calculate size of a number ofter JSON serialization.
 *
 * @param value - Number value to calculate size.
 * @returns Number of bytes used to store whole number in JSON.
 */
export function calculateNumberSize(value: number): number {
  if (value === 0) {
    return JsonSize.ZERO;
  }

  let size = 0;
  // Work with absolute (positive) numbers only
  let absNum = value;
  if (value < 0) {
    absNum = Math.abs(value);
    // Initially, count on a sign of a number (-)
    size += 1;
  }

  // If absolute value of a number is greater than 1 and is a whole number,
  // then perform the fastest operation to calculate its size
  // Portion of numbers passed to this function will fall into this category,
  // so it will not be required to do to string conversion and manipulation.
  if (absNum - Math.floor(absNum) === 0) {
    size += Math.floor(Math.log10(absNum) + 1);
    return size;
  }

  // Work with decimal numbers
  // First calculate the size of the whole part of a number
  if (absNum >= 1) {
    size += Math.floor(Math.log10(absNum) + 1);
  } else {
    // Because absolute value is less than 1, count size of a zero digit
    size += JsonSize.ZERO;
  }
  // Then add the dot '.' size
  size += JsonSize.DOT;
  // Then calculate the number of decimal places
  size += getNumberOfDecimals(absNum);

  return size;
}

/**
 * Calculate the number of decimals for a given number.
 *
 * @param value - Decimal number.
 * @returns Number of decimals.
 */
export function getNumberOfDecimals(value: number): number {
  if (Math.floor(value) === value) {
    return 0;
  }

  const str = Math.abs(value).toString();

  if (str.indexOf('e-') > -1) {
    return parseInt(str.split('e-')[1], 10);
  }

  return str.split('.')[1].length;
}
```

Here you can see how I explained the cases in which Math.log10 was not working well and how I handled that. But, that was alright to overcome in this way somehow I believe. By looking at it in a probabilistic way some numbers would fall into each categories and would consume different amount of the resources (memory and time) which looks good to me. But, at the end I ended up in the need to check for exponential notations initially which I was only able to do by parsing the number, which makes all of this redundant then.

Maybe I'm overlooking something or missing some ideas. Let me know if you have some I'll be happy to try something like this again!

Reply:
Are numbers just stringified when being JSON serialized? I.e. does it keep the exponential notation?

Reply:
Yes, it keeps the notation. And also JavaScript has it own ways for dealing with specific numbers that are smaller than some value or larger than some value.

If you look at this example, you'll see how it works:

```javascript
const objectToSerialize = {
    verySmallNumber: 0.00000000005,
    exponentialNumber: -5e-11,
    someOtherExponentialNumber: 3e+15,
    justAnotherExponentialNumber: 3e+21
}

console.log(objectToSerialize);
console.log(JSON.stringify(objectToSerialize));
```
Will output this:
```javascript
{
  verySmallNumber: 5e-11,
  exponentialNumber: -5e-11,
  someOtherExponentialNumber: 3000000000000000,
  justAnotherExponentialNumber: 3e+21
}
```
```json
{"verySmallNumber":5e-11,"exponentialNumber":-5e-11,"someOtherExponentialNumber":3000000000000000,"justAnotherExponentialNumber":3e+21}
```